drive/uniquefolder/uniquefolder.html

```diff
@@ -0,0 +1,194 @@
+       * folder. If it doesn't exist, check trash. If it isn't trashed, create it.
+       * Note that even though the folder has a unique key and name, its not impossible
+       * to have two folderes with that key and name.
+       *
+       * 
+       */
+      function verifyFolder() {
+        // Notice we use the query string parameter (q) to identify the criteria for folders to return. One criteria is that
+        // the folder must have appProperties with a specific key of "myfolderkey" and value of "1234" set.  
+        //
+        // By default, files.list in v3 only returns a subset of all fields, such as file id. If you want specific fields returned,
+        // you must request them using the fields parameter. In the following example, we ask for the trashed field to be returned.
+        // By defining the fields parameter, only the fields defined by that parameter are returned (file id would no longer be
+        // returned).
+        //
+        // Note that you have to ask for specific fields to be returned and those fields are within the context of the container
+        // field's name which is "files," such as files/trashed, files/id.
+        gapi.client.drive.files.list({
+          'q': '"me" in owners and mimeType="application/vnd.google-apps.folder" and appProperties has {key="myfolderkey" and value="1234"} and visibility="limited"',
+          'fields': 'files/trashed, files/id'
+        }).then(function(response) {
+          files = response.result.files;
+          if (files && files.length > 0) {
+            appendPre(files.length + ' folder(s) found. Checking trashed status');
+            // Cycle through all files and check if they are trashed.
+            for (filesCounter=0; filesCounter < files.length; filesCounter++){
+              checkTrashed();
+            }
+          } else {
+            appendPre('Folder not found. Creating folder');
+            createFolder();
+          }
+        });
+      }
+ 
+      /**
+       * Identify whether a folder is trashed. If the folder is trashed, call untrash().
+       */
+      function checkTrashed() {
+        if (files[filesCounter].trashed == true) {
+          appendPre('Folder ' + (filesCounter+1) + ' is trashed. Untrash.');
+          untrash();
+        } else {
+          appendPre('Folder ' + (filesCounter+1) + ' is not trashed.');
+        }	
+      }
+
+      /**
+       * Untrash a folder.
+       */
+      function untrash() {
+        fileId = files[filesCounter].id;
+        appendPre(filesCounter);
+        // We set the trashed parameter to false to untrash the folder.
+        gapi.client.drive.files.update({
+          'fileId': 'fileId',
+          'trashed': false
+        }).then(function(response) {
+          appendPre('Folder ' + filesCounter + ' is untrashed.');
+        });
+      }
+
+      /**
+       * Create a folder with appProperties including a key of "myfolderkey" and a value of "1234". The key and value are used to
```

Comment:
Don't understand your second sentence. Are you saying that appProperties isn't needed when you use drive.file and crate only one file. Are you saying that drive.file only provides access to files created by that app?

Honestly, as I went on with this example, I wasn't sure it really told a good story for when to use appProperties.

Are you suggesting I change this sample such that appProperties ensures the foldler is in the root of a myDrive (isApplicationRoot=true)?

Reply:
Correct, when using the file scope drive.files.list only returns the files that the app has access to (created or user picks via 'open with...' or picker. As this sample stands, the app properties don't add any value. It's also not clear what value was *intended* as the key/value doesn't convey any particular meaning. I don't know what myfolderkey means or why the value 1234 is used.

So either:
1) Change the key/value pair to more accurately reflect the intention.
2) Drop the app properties from this sample.

The thing about isApplicationRoot was just an example of a *potential* story for them. I could see wanting to disambiguate between the root/first folder of the app and subfolders that may be created and returned in files.list, and app properties are one way to accomplish that. Another use case would be something like a fax app having inbox/outbox. App properties can be used to tag a folder with the 'role' it has so that you don't have to rely on storing the IDs separately or rely on naming conventions.